### PR TITLE
Upload contract icon

### PR DESCRIPTION
**Pull request by matjazonline**

### CODE REVIEW

**Comment by matjazonline** [comment]
using other, newer implementation

File: src/controllers/updateTokenIcon.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+      res.status(403).json({error: "Same file is being uploaded"})
+    }
+  } catch (err: any) {
+    console.error(err);
+  }
+
+  //calculating the hash for verification
+  const calculatedHash = generateSHA256Hash(file);
+  //validating the hash of file
+  if (calculatedHash != fileHash) {
```
Comment (boo-0x) [inline src/controllers/updateTokenIcon.ts:84]: There is no need to receive the file hash, as we already calculate it here

File: src/controllers/updateTokenIcon.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  //file is same - now checking if the icon is already uploaded or not
+  // to make it fast - checking it in localdb
+  if(verifiedBackup?.contractData.iconUrl){
+    res.status(403).json({error: "icon url already present"});
+  } 
+
+  // address of user who is trying to upload icon
+  const signerAddress = ethers.utils.verifyMessage(JSON.stringify(req.body.signMsg), ethers.utils.toUtf8Bytes(req.body.signature));
```
Comment (boo-0x) [inline src/controllers/updateTokenIcon.ts:101]: We have to use Polkadot.js for Reef signatures.

Also, we have to check for a specific message and not only for the signer address, as someone could send the signature of another user. So the flow would be:
- Front end asks for the nonce of a certain signer address.
- We generate a random nonce, store it in DB and return it to the front end.
- User signs a message with the nonce and adds it in the request.
- We check that the signature is signing message with nonce. 


See example: 
https://github.com/sqwid-app/sqwid-backend/blob/292e8f1b2f4e6396086178648c77f91df0400f49/lib/verify.js#L7
https://github.com/sqwid-app/sqwid-backend/blob/292e8f1b2f4e6396086178648c77f91df0400f49/routes/auth.js#L8

File: src/controllers/updateTokenIcon.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+
+  if (signerAddress === contract?.contractData.signer.id) {
```
Comment (boo-0x) [inline src/controllers/updateTokenIcon.ts:104]: To check if `signerAddress` can update the icon we have to:

1) Check if the contract has `iconUri` field: Using Ethers, we call the `iconUri()` function of the contract (use a try-catch as the function might not exist) to check if the value exists on chain. If it exists, we do not allow to update the value.

2) Check if `signerAddress` is the owner of the contract: Using Ethers, we call the `owner()` function of the contract (use a try-catch as the function might not exist) and if it has value, get the associated evmAddress of `signerAddress` and check if it matches the address returned by `owner()`. If they do not match, we do not allow to update the value.

3) Check if `signerAddress` is the deployer of the contract: `if (signerAddress === contract.contract.signer.id)`.

File: src/controllers/updateTokenIcon.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    // Add a new record to the local db with a pending status
+    await uploadIconRepository.create({
+    address: contractAddress,
+    filesize: file.size,
+    pending:true,
+    });
+
+    //obtain the ipfsHash by uploading the file
+    const ipfsHash = upload(file);
+    await updateVerifiedContractData(contractAddress, 'ipfs://' + ipfsHash);
```
Comment (boo-0x) [inline src/controllers/updateTokenIcon.ts:114]: The second param should be a json with this format: { "iconUrl": `ipfs://'${ipfsHash}` }

File: src/services/updateTokenIcon.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  }
+
+export const upload =  async (file)=>{   
+    const formData = new FormData()
+    const projectId = process.env.IPFS_PROJECT_ID
+    const projectSecret = process.env.IPFS_PROJECT_SECRET
+    formData.append('file', file)
+    try {
+      const response = await axios.post(
+        'https://ipfs.infura.io:5001/api/v0/add',
```
Comment (boo-0x) [inline src/services/updateTokenIcon.ts:15]: We can add this as an env variable.

**Review by boo-0x — CHANGES_REQUESTED** [review]
See comments in code.

**Pull request closed without merging**